src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Extern.java

```diff
@@ -223,6 +226,134 @@ public boolean link(String url, String elemlisturl, Reporter reporter) throws Do
+            if (docUrl == null) {
+                return;
+            }
+            DocPath elementListPath = getPlatformElementList(versionNumber);
+            URL elementListUrl = AbstractDoclet.class.getResource(elementListPath.getPath());
+            if (elementListUrl == null) {
+                reporter.print(Kind.WARNING, resources.getText("doclet.Resource_error", elementListPath.getPath()));
+            } else {
+                try {
+                    readListFromURL(docUrl, elementListUrl);
+                } catch (IOException exc) {
+                    throw new Fault(resources.getText(
+                            "doclet.Resource_error", elementListPath.getPath()), exc);
+                }
+            }
+        } catch (Fault f) {
+            reporter.print(Kind.ERROR, f.getMessage());
+        }
+    }
+
+    /**
+     * Return the resource path for the package or element list for the given {@code version}.
+     * @param version the platform version number
+     * @return the resource path
+     */
+    private DocPath getPlatformElementList(int version) {
+        String filename = version <= 9
+                ? "package-list-" + version + ".txt"
+                : "element-list-" + version + ".txt";
+        return DocPaths.RESOURCES.resolve("releases").resolve(filename);
+    }
+
+    /**
+     * Return the default URL for the platform API documentation for the given {@code version}.
+     * @param version the platform version number
+     * @return the URL as String
+     */
+    private String getDefaultPlatformDocs(int version) {
+        Resources resources = configuration.getDocResources();
+        return version <= 10
+                ? resources.getText("doclet.platform.docs.old", version)
+                : isPrerelease(version)
+                    ? resources.getText("doclet.platform.docs.ea", version)
+                    : resources.getText("doclet.platform.docs.new", version);
+    }
+
+    /**
+     * Retrieve and return the custom URL for the platform API documentation for the given
+     * {@code version} from the properties file at {@code linkPlatformProperties}.
+     * @param version the platform version number
+     * @param linkPlatformProperties path pointing to a properties file
+     * @return the custom URL as String
+     */
+    private String getCustomPlatformDocs(int version, String linkPlatformProperties) throws Fault {
+        String url;
+        try {
+            Properties props = new Properties();
+            InputStream inputStream;
+            if (isUrl(linkPlatformProperties)) {
+                inputStream = toURL(linkPlatformProperties).openStream();
+            } else {
+                inputStream = DocFile.createFileForInput(configuration, linkPlatformProperties).openInputStream();
+            }
+            try (inputStream) {
+                props.load(inputStream);
+            }
+            url = props.getProperty("doclet.platform.docs." + version);
```

Comment:
Similar to other file-or-url arguments: good!

Reply:
As a possibly-later cleanup, should we have a single utility method somewhere (in this class) to open a stream on a file-or-url?